src/security/builtin_plugins/authentication/src/auth_utils.c

```diff
@@ -0,0 +1,1530 @@
+                    char *msg = get_openssl_error_message();
+                    result = DDS_SECURITY_VALIDATION_FAILED;
+                    DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result, "RSA key not correct : %s", msg);
+                    ddsrt_free(msg);
+                }
+            }
+            RSA_free(rsaKey);
+        }
+        break;
+    case EVP_PKEY_EC:
+        if (EVP_PKEY_bits(key) != 256) {
+            result = DDS_SECURITY_VALIDATION_FAILED;
+            DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result, "EC %s has unsupported key size (%d)", sub, EVP_PKEY_bits(key));
+        } else {
+            EC_KEY *ecKey = EVP_PKEY_get1_EC_KEY(key);
+            if (ecKey) {
+                if (EC_KEY_check_key(ecKey) != 1) {
+                    char *msg = get_openssl_error_message();
+                    result = DDS_SECURITY_VALIDATION_FAILED;
+                    DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result, "EC key not correct : %s", msg);
+                    ddsrt_free(msg);
+                }
+            }
+            EC_KEY_free(ecKey);
+        }
+        break;
+    default:
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result, "%s has not supported type", sub);
+        break;
+    }
+
+    return result;
+}
+
+static DDS_Security_ValidationResult_t
+check_certificate_type_and_size(
+    X509 *cert,
+    DDS_Security_SecurityException *ex)
+{
+    EVP_PKEY *pkey;
+    DDS_Security_ValidationResult_t result = DDS_SECURITY_VALIDATION_OK;
+
+    assert(cert);
+
+    pkey = X509_get_pubkey(cert);
+    if (pkey) {
+        result = check_key_type_and_size(pkey, false, ex);
+    } else {
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result, "X509_get_pubkey failed");
+    }
+    EVP_PKEY_free(pkey);
+
+    return result;
+}
+
+DDS_Security_ValidationResult_t
+check_certificate_expiry(
+    X509 *cert,
+    DDS_Security_SecurityException *ex)
+{
+    DDS_Security_ValidationResult_t result = DDS_SECURITY_VALIDATION_OK;
+
+    assert(cert);
+
+    if( X509_cmp_current_time(X509_get_notBefore( cert )) == 0 ){
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_CERT_STARTDATE_INVALID_CODE, (int)result, DDS_SECURITY_ERR_CERT_STARTDATE_INVALID_MESSAGE);
+
+    }
+    if( X509_cmp_current_time(X509_get_notAfter( cert )) == 0 ){
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_CERT_EXPIRED_CODE, (int)result, DDS_SECURITY_ERR_CERT_STARTDATE_INVALID_MESSAGE);
```

Comment:
Presumably start times precede end times, and so both errors cannot occur simultaneously (but one ought to check). But the current time could be reset in between the two calls, so you're potentially overwriting the start date error, and the ``DDS_SECURITY_ERR_CERT_STARTDATE_INVALID_MESSAGE`` here seems out of place.

Reply:
This command has generated a certificate with a start date later than the end date:
`openssl x509 -req -sha256 -days -365   -in server.csr -signkey server.key -out server.crt`
Result:
>         Validity
>             Not Before: Oct 17 12:39:42 2019 GMT
>             Not After : Oct 17 12:39:42 2018 GMT

Thanks for `DDS_SECURITY_ERR_CERT_STARTDATE_INVALID_MESSAGE `